lib/nostrum/struct/invite.ex

```diff
@@ -3,74 +3,93 @@ defmodule Nostrum.Struct.Invite do
   Struct representing a Discord invite.
   """
 
-  alias Nostrum.Struct.Invite.{Channel, Guild}
+  alias Nostrum.Struct.Channel
+  alias Nostrum.Struct.Guild
+  alias Nostrum.Struct.Invite.Metadata
   alias Nostrum.Struct.User
+  alias Nostrum.Util
 
-  @typedoc "Invite code"
+  defstruct [
+    :code,
+    :guild,
+    :channel,
+    :inviter,
+    :metadata
+  ]
+
+  @typedoc """
+  The invite code (unique ID).
+  """
   @type code :: String.t()
 
-  @typedoc "Guild the invite is for"
+  @typedoc """
+  The guild this invite is for.
+  """
   @type guild :: Guild.t()
 
-  @typedoc "Channel the invite is for"
+  @typedoc """
+  The channel this invite is for.
+  """
   @type channel :: Channel.t()
 
-  @typedoc "User who created the invite"
+  @typedoc """
+  The user who created this invite.
+  """
   @type inviter :: User.t()
 
-  @typedoc "Number of times this invite has been used"
-  @type uses :: integer
-
-  @typedoc "Number of times this invite can be used"
-  @type max_uses :: integer
+  @typedoc """
+  The extra metadata this invite contains.
 
-  @typedoc "Duration in seconds after which the invite expires"
-  @type max_age :: integer
-
-  @typedoc "Whether the invite gives temporary membership"
-  @type temporary :: boolean
-
-  @typedoc "When the invite was created"
-  @type created_at :: String.t()
+  This field is `nil` unless this invite is returned by
+  `Nostrum.Api.get_channel_invites/1`.
+  """
+  @type metadata :: Metadata.t()
```

Comment:
Sorry, but I think I got a little confused by the chat we had in the Discord channel.

Do we want to break away from the Discord docs on this one? I do agree that we'll have a bunch of nils in one of our Invite types, but that's what we've been doing.

What are you thoughts on this?

Reply:
You are right. We probably should stick to the original layout we've been doing for sometime now. I'll be revising my spec and design.

However, hear me out on a concern of mine. My thoughts right now are that we're going to have to eventually break from the API. Especially when we start building structs for the Gateway events. The gateway events alter all sorts of JSON schema for its payloads. For example, some of the Gateway events contain an extra `:guild_id` field. Because the discord API has so many weird rules that bend out a struct's schema, I was wondering if we should break away from modeling a perfect 1 to 1 with the discord API. Maybe gateway structs should be an exception to the design so far? Let me know your thoughts while I wrap this PR up. 😈

Reply:
I think there's a couple different ways we could handle it.

1. We could just make new structs for the gw payloads. This would be a lot of rewriting, but would follow suit with what we've been doing.

2. We could forgo gateway structs in favor of another format, like a tuple or something. Since the user eventually handles these events as a tuple, I think this would be sane. For something like an added `guild_id` we could supply a tuple like `{:TYPING_START, guild_id :: Snowflake.t(), member :: Member.t()}`. Downside is that there would be some oddities like members inside message creates, as we probably don't want users to have to have `{:MESSAGE_CREATE, member :: Member.t(), message :: Message.t()}`.

3. We could expose these new fields in their respective structs. This has the downside of weird keys in our structs, but we've handled this before by just have different types where we explicitly state something will be `nil`. 

I think I'm leaning towards a combination of 1 & 2 so we could have something like

```elixir
%MessageCreateStruct{
  message: Message.to_struct(payload.msg),
  member: Member.to_struct(payload.member)
}
 ```

The downside here is that we'd have to redo all of the types for the consumer, but we're doing that anyway. I think this would be nice so we don't have to send nested tuples, and it'd be nice to have something like

```elixir
%UserUpdateStruct{
  old_user: old_user_struct,
  new_user: new_user_struct
}
```

Let me know your thoughts, maybe we should look more at actual cases of this to better make our choice.